spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,301 @@
+  CloseableIterable<IndexedTask> getChangesWithRateLimit(long snapshotId, int index, boolean isStarting, int maxFiles) {
+    List<CloseableIterable<IndexedTask>> snapshotTasks = Lists.newArrayList();
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+    int currentFiles = 0;
+
+    if (isStarting) {
+      CloseableIterable<FileScanTask> iter = buildTableScan()
+          .option("use-worker-pool", "false")
+          .useSnapshot(snapshotId)
+          .planFiles();
+      CloseableIterable<IndexedTask> tasks = filterTasks(indexTasks(iter, snapshotId), snapshotId, index);
+      Pair<Integer, CloseableIterable<IndexedTask>> limitedTasks = limitTasks(tasks, currentFiles, maxFiles);
+      currentFiles = limitedTasks.getLeft();
+      snapshotTasks.add(limitedTasks.getRight());
+
+      if (currentFiles >= maxFiles) {
+        // If the files to be processed already reach to max-files-batch, will not continue to plan files.
+        return CloseableIterable.concat(snapshotTasks);
+      }
+    }
+
+    Long start = isStarting ? Long.valueOf(snapshotId) : table.snapshot(snapshotId).parentId();
+    Preconditions.checkState(start != null, "Start snapshot id should exist");
+    ImmutableList<Long> snapshotIds = ImmutableList.<Long>builder()
+        .addAll(SnapshotUtil.snapshotIdsBetween(table, start, currentSnapshotId))
+        .add(start)
+        .build()
+        .reverse();
+
+    for (int i = 0; i < snapshotIds.size() - 1; i++) {
+      long fromSnapshotId = snapshotIds.get(i);
+      long toSnapshotId = snapshotIds.get(i + 1);
+
+      CloseableIterable<FileScanTask> iter = buildTableScan()
+          .option("use-worker-pool", "false")
+          .appendsBetween(fromSnapshotId, toSnapshotId)
+          .planFiles();
+      CloseableIterable<IndexedTask> tasks = filterTasks(indexTasks(iter, toSnapshotId), snapshotId, index);
+      Pair<Integer, CloseableIterable<IndexedTask>> limitedTasks = limitTasks(tasks, currentFiles, maxFiles);
+      currentFiles = limitedTasks.getLeft();
+      snapshotTasks.add(limitedTasks.getRight());
+
+      if (currentFiles >= maxFiles) {
+        break;
+      }
+    }
+
+    return CloseableIterable.concat(snapshotTasks);
+  }
+
+  private static CloseableIterable<IndexedTask> indexTasks(CloseableIterable<FileScanTask> iter, long snapshotId) {
+    return CloseableIterable.transformWithIndex(iter, (idx, task) -> new IndexedTask(snapshotId, task, idx.intValue()));
+  }
+
+  private static CloseableIterable<IndexedTask> filterTasks(CloseableIterable<IndexedTask> iter,
+                                                            long startSnapshotId, int index) {
+    return CloseableIterable.filter(iter, t -> t.snapshotId() != startSnapshotId || t.index() > index);
+  }
+
+  private static Pair<Integer, CloseableIterable<IndexedTask>> limitTasks(CloseableIterable<IndexedTask> iter,
+                                                                          int currentFiles, int maxFiles) {
```

Comment:
This will scan metadata files multiple times. Using `Iterables.getLast` will require scanning manifests, but then the Iterable returned will also need to re-do that work.

I think the right solution is to turn the table state into a concatenated `CloseableIterable` like you're doing, but then actually consume from an `iterator` from that result. Then this would return a pair of `List<FileScanTask>` and `Iterator<IndexedTask>`. The list is the next batch's tasks, and the iterator is re-used.